### ISSUE TITLE
Trained YOLOv8 model converted to CoreML doesn't give any predictions

### ISSUE THREAD

**Issue by tardoe** [ultralytics/ultralytics#13794]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I'm having some issues exporting a YOLOv8 model I trained to CoreML. Testing the model in yolo format predicts just fine, after converting to CoreML, testing with coremltools predictions and using the Xcode preview return no results. 

I trained the custom model based on 640x640 jpg images across 43 classes, used the yolov8m model as the base, using CUDA on an RTX4090. The training was done like this: `results = model.train(data="yolov8_dataset/dataset.yaml", epochs=350, device=0)`. The training ended normally after 330 epochs having not changed in 100 epochs. 

This is how I'm exporting the model: `model.export(format="coreml", nms=True, half=False, imgsz=640)` using coremltools 7.2.
I'm using  `YOLOv8.2.35 / Python-3.11.6 torch-2.2.0 CPU (Apple M1 Max)` to do the conversion. No errors in the conversion process 

Testing against the designated 640x640 test images from my original dataset.

I also tried this same process on the YOLOv8m "vanilla" model and it worked just fine, the Xcode preview showed predictions - what have I likely done wrong with the custom model? 

What am I likely missing here?

**Comment by glenn-jocher** [comment]
Hi @tardoe,

Thank you for reaching out and providing detailed information about your issue. Let's work through this together to identify the problem.

Firstly, it's great to hear that the vanilla YOLOv8m model works fine after conversion to CoreML. This indicates that the conversion process itself is functioning correctly. The issue seems to be specific to your custom-trained model.

Here are a few steps to help diagnose and potentially resolve the issue:

1. **Verify Model Performance Before Export**:
   Ensure that your custom-trained model performs as expected in its original format before conversion. You can do this by running predictions on your test images using the `predict` method in the YOLO format:
   ```python
   results = model.predict(source='path/to/test/images', imgsz=640)
   results.show()
   ```

2. **Check CoreML Export Parameters**:
   Your export command looks correct, but let's ensure that the parameters are optimal. You might want to try exporting without the `nms` parameter to see if it affects the predictions:
   ```python
   model.export(format="coreml", half=False, imgsz=640)
   ```

3. **Inspect CoreML Model in Xcode**:
   After exporting, use Xcode to inspect the CoreML model. Ensure that the input and output shapes match your expectations. Sometimes, discrepancies in input/output shapes can lead to no predictions.

4. **Test CoreML Model with CoreMLTools**:
   Use `coremltools` to test the CoreML model directly in Python before deploying it to Xcode. This can help isolate whether the issue is with the model or the deployment environment:
   ```python
   import coremltools as ct
   from PIL import Image
   import numpy as np

   # Load the CoreML model
   coreml_model = ct.models.MLModel('yolov8n.mlpackage')

   # Load and preprocess an image
   image = Image.open('path/to/test/image.jpg').resize((640, 640))
   image_np = np.array(image).astype(np.float32)

   # Make a prediction
   prediction = coreml_model.predict({'image': image_np})
   print(prediction)
   ```

5. **Ensure Compatibility**:
   Verify that the versions of `torch`, `ultralytics`, and `coremltools` are compatible. Sometimes, version mismatches can cause unexpected behavior. Ensure you are using the latest versions:
   ```bash
   pip install --upgrade torch ultralytics coremltools
   ```

If the issue persists, please provide a minimum reproducible example of your code and dataset configuration. This will help us reproduce the bug and investigate further. You can find guidelines for creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out with any additional questions or updates on your progress. We're here to help!

**Comment by tardoe** [comment]
Thanks for the advice, I'll address each one accordingly:

1. The model performs as expected, I saw accurate predictions for 95% of of 225 image testing image set (separate to the training set).
2. Exporting without NMS means the import into Xcode doesn't show the class labels and doesn't allow me to test the prediction inside Xcode.
3. Testing with coremltools results in the following error:
```python
>>> import coremltools as ct
>>> from PIL import Image
>>> import numpy as np
>>> coreml_model = ct.models.MLModel('./runs/detect/train3/weights/best.mlpackage')
>>> image = Image.open('./dataset/test/images/test_21.jpg')
>>> image_np = np.array(image).astype(np.float32)
>>> prediction = coreml_model.predict({'image': image_np})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "./venv/lib/python3.11/site-packages/coremltools/models/model.py", line 627, in predict
    return MLModel._get_predictions(self.__proxy__, verify_and_convert_input_dict, data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "./venv/lib/python3.11/site-packages/coremltools/models/model.py", line 669, in _get_predictions
    preprocess_method(data)
  File "./venv/lib/python3.11/site-packages/coremltools/models/model.py", line 615, in verify_and_convert_input_dict
    self._verify_input_dict(d)
  File "./venv/lib/python3.11/site-packages/coremltools/models/model.py", line 741, in _verify_input_dict
    self._verify_pil_image_modes(input_dict)
  File "./venv/lib/python3.11/site-packages/coremltools/models/model.py", line 752, in _verify_pil_image_modes
    raise TypeError(msg.format(input_desc.name))
TypeError: Image input, 'image' must be of type PIL.Image.Image in the input dict
```
Trying again, by just passing in the PIL Image object, I get a valid prediction:
```python
>>> prediction = coreml_model.predict({'image': image})
>>> print(prediction)
{'var_1140': array([[[     9.2656,      40.438,      31.188, ...,         501,         523,       543.5],
        [     33.344,      4.4141,      4.3789, ...,         621,         618,         610],
        [     23.406,      92.625,      166.38, ...,         398,       373.5,         249],
        ...,
        [          0,           0,           0, ...,           0,           0,           0],
        [          0,           0,           0, ...,           0,           0,           0],
        [          0,           0,           0, ...,           0,           0,           0]]], dtype=float32)}
```
Though I don't understand the output. This is on the model converted *without* `nms=True`.
If I export the model with nms , I don't get a result:
```python
>>> from PIL import Image
>>> import numpy as np
>>> coreml_model = ct.models.MLModel('cpe_id/cpe_id_2/runs/detect/train3/weights/best.mlpackage')
>>> image = Image.open('/Users/tim/Development/cpe-id/cpe_id_model/cpe_id/cpe_id_2/dataset/test/images/tplink_vx220g2v_21.jpg')
>>> prediction = coreml_model.predict({'image': image})
>>> prediction
{'coordinates': array([], shape=(0, 4), dtype=float32), 'confidence': array([], shape=(0, 44), dtype=float32)}
```
However, trying to use the model without `nms=True` in Xcode throws an error because the labels are missing from the metadata, how do you fix that?

4. I'm running the latest versions of all tools:
```bash
❯ pip freeze | grep -E "torch=|ultralytics=|coremltools="
coremltools==7.2
torch==2.2.0
ultralytics==8.2.36
```

**Comment by glenn-jocher** [comment]
Hi @tardoe,

Thank you for the detailed follow-up! Let's address each point to help you resolve the issue:

1. **Model Performance**:
   It's great to hear that your model performs well on the test set. This confirms that the training process was successful.

2. **Exporting Without NMS**:
   Since exporting without NMS results in missing class labels in Xcode, it's essential to include NMS during export. This ensures that the model's predictions are properly formatted for Xcode.

3. **CoreMLTools Testing**:
   The error you encountered with `coremltools` indicates that the input should be a `PIL.Image.Image` object. Your subsequent test with the PIL image object worked, but the output format seems unclear. When exporting with `nms=True`, the model should include the necessary metadata for proper predictions. Let's ensure the export process is correct:
   ```python
   model.export(format="coreml", nms=True, half=False, imgsz=640)
   ```

   If the exported model with `nms=True` still doesn't give results, it might be due to the input preprocessing or post-processing steps. Ensure that the input image is correctly preprocessed and matches the expected input format of the CoreML model.

4. **CoreML Model Output**:
   The output you received without NMS (`'var_1140': array([...])`) seems to be raw predictions. When NMS is applied, the output should be more structured, including coordinates and confidence scores. If the model with `nms=True` returns empty arrays, it might be due to a thresholding issue. Ensure that the confidence threshold is set appropriately during inference.

5. **Metadata for Xcode**:
   To fix the issue with missing labels in Xcode, ensure that the CoreML model includes the class labels in its metadata. You can manually add metadata to the CoreML model using `coremltools`:
   ```python
   import coremltools as ct

   # Load the CoreML model
   coreml_model = ct.models.MLModel('path/to/your/model.mlpackage')

   # Add class labels to the model's metadata
   coreml_model.user_defined_metadata['classes'] = 'path/to/your/classes.txt'

   # Save the updated model
   coreml_model.save('path/to/your/updated_model.mlpackage')
   ```

6. **Versions**:
   You're using the latest versions of `torch`, `ultralytics`, and `coremltools`, which is excellent. This ensures compatibility and access to the latest features and fixes.

If the issue persists, please provide a minimum reproducible example of your code and dataset configuration. This will help us reproduce the bug and investigate further. You can find guidelines for creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out with any additional questions or updates on your progress. We're here to help! 😊

**Comment by tardoe** [comment]
Thanks, for the suggestions.

1. Agreed.
2. Noted
3. When testing the model in Xcode, I use the same, correctly formatted images as I had in the testing set. They match the model's expected input. 
4. The confidence thresholds in the model are already very low:
```json
  userDefined {
    key: "Confidence threshold"
    value: "0.25"
  }
  userDefined {
    key: "IoU threshold"
    value: "0.45"
  }
```
5. The missing labels are only the case when nms=False which is expect as per your comments in 2.
6.  Ok.

I believe I've already provided the minimal examples above for reproducibility.

**Comment by glenn-jocher** [comment]
Hi @tardoe,

Thank you for the detailed follow-up and for confirming the steps you've taken so far. Let's dive deeper into resolving this issue.

### Key Points Recap:
1. **Model Performance**: Confirmed to be accurate before export.
2. **Export with NMS**: Necessary for proper predictions in Xcode.
3. **Testing in Xcode**: Using correctly formatted images.
4. **Confidence Thresholds**: Already set to low values.
5. **Labels with NMS**: Confirmed to be included when `nms=True`.
6. **Versions**: Using the latest versions of all tools.

### Next Steps:
Given that the CoreML model with `nms=True` still returns empty predictions, let's ensure the input preprocessing and post-processing steps are correctly handled. Here are a few additional checks and steps:

1. **Verify Input Preprocessing**:
   Ensure the input image is preprocessed to match the model's expected format. This includes resizing and normalizing the image:
   ```python
   from PIL import Image
   import numpy as np
   import coremltools as ct

   # Load the CoreML model
   coreml_model = ct.models.MLModel('path/to/your/model.mlpackage')

   # Load and preprocess the image
   image = Image.open('path/to/test/image.jpg').resize((640, 640))
   image_np = np.array(image).astype(np.float32) / 255.0  # Normalize if required

   # Make a prediction
   prediction = coreml_model.predict({'image': image_np})
   print(prediction)
   ```

2. **Check Post-Processing**:
   Ensure that the post-processing steps, such as applying the confidence and IoU thresholds, are correctly implemented in your CoreML model. This can sometimes be a source of issues if not properly configured.

3. **Inspect Model Metadata**:
   Verify that the model metadata includes the correct class labels and other necessary information:
   ```python
   import coremltools as ct

   # Load the CoreML model
   coreml_model = ct.models.MLModel('path/to/your/model.mlpackage')

   # Inspect metadata
   print(coreml_model.user_defined_metadata)
   ```

4. **Test with Different Confidence Thresholds**:
   Try adjusting the confidence threshold to see if it affects the predictions:
   ```python
   coreml_model.user_defined_metadata['Confidence threshold'] = '0.1'
   coreml_model.save('path/to/your/updated_model.mlpackage')
   ```

If the issue persists, please provide a more detailed minimum reproducible example, including the specific steps and code used for training, exporting, and testing the model. This will help us reproduce the issue on our end and investigate further. You can refer to our [minimum reproducible example guide](https://docs.ultralytics.com/help/minimum_reproducible_example) for more details.

Thank you for your patience and cooperation. We're here to help you get this resolved! 😊

**Comment by tardoe** [comment]
Thanks,

1. The process you provided isn't valid, you can't pass a numpy array to coremltools.model.predict()
2. As far as I know, these are correct in the exported metadata.
3. This is confirmed correct
4. Setting a low confidence of 0.01 does offer any predictions.

**Comment by glenn-jocher** [comment]
Hi @tardoe,

Thank you for your detailed feedback and for confirming the steps you've taken. Let's address your points to help resolve the issue:

1. **CoreMLTools Prediction**:
   You're right; CoreMLTools requires a `PIL.Image.Image` object for prediction. Apologies for the confusion. Here's the correct approach:
   ```python
   from PIL import Image
   import coremltools as ct

   # Load the CoreML model
   coreml_model = ct.models.MLModel('path/to/your/model.mlpackage')

   # Load and preprocess the image
   image = Image.open('path/to/test/image.jpg').resize((640, 640))

   # Make a prediction
   prediction = coreml_model.predict({'image': image})
   print(prediction)
   ```

2. **Metadata Verification**:
   It's good to hear that the metadata is correct. This ensures that the model should theoretically work as expected.

3. **Input Format**:
   Confirmed to be correct, which is crucial for accurate predictions.

4. **Low Confidence Threshold**:
   If setting a low confidence threshold still doesn't yield predictions, it might indicate an issue with the model's post-processing or the way predictions are handled in CoreML.

### Next Steps:
To further diagnose the issue, could you please provide a minimum reproducible example of your code and dataset configuration? This will help us reproduce the bug and investigate a solution. You can find guidelines for creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Additionally, please ensure you are using the latest versions of `torch`, `ultralytics`, and `coremltools`. Sometimes, updates can resolve unexpected issues.

Thank you for your cooperation and patience. We're here to help you get this resolved! 😊

**Comment by tardoe** [comment]
1. I've provided this example above that illustrates the exact issue: 
```python
>>> from PIL import Image
>>> import numpy as np
>>> coreml_model = ct.models.MLModel('cpe_id/cpe_id_2/runs/detect/train3/weights/best.mlpackage')
>>> image = Image.open('/Users/tim/Development/cpe-id/cpe_id_model/cpe_id/cpe_id_2/dataset/test/images/tplink_vx220g2v_21.jpg')
>>> prediction = coreml_model.predict({'image': image})
>>> prediction
{'coordinates': array([], shape=(0, 4), dtype=float32), 'confidence': array([], shape=(0, 44), dtype=float32)}
```

**Comment by glenn-jocher** [comment]
Hi @tardoe,

Thank you for providing the code example and detailed information. Let's work through this together to resolve the issue.

### Steps to Diagnose and Resolve:

1. **Verify Latest Versions**:
   Ensure you are using the latest versions of `torch`, `ultralytics`, and `coremltools`. Sometimes, updates can resolve unexpected issues. You can upgrade your packages using:
   ```bash
   pip install --upgrade torch ultralytics coremltools
   ```

2. **Check Model Export**:
   Ensure the model is exported correctly with NMS enabled. This is crucial for proper predictions in CoreML:
   ```python
   model.export(format="coreml", nms=True, half=False, imgsz=640)
   ```

3. **CoreML Prediction**:
   Use the following code to test predictions with CoreMLTools:
   ```python
   from PIL import Image
   import coremltools as ct

   # Load the CoreML model
   coreml_model = ct.models.MLModel('path/to/your/model.mlpackage')

   # Load and preprocess the image
   image = Image.open('path/to/test/image.jpg').resize((640, 640))

   # Make a prediction
   prediction = coreml_model.predict({'image': image})
   print(prediction)
   ```

4. **Inspect Metadata**:
   Verify that the model metadata includes the correct class labels and other necessary information:
   ```python
   import coremltools as ct

   # Load the CoreML model
   coreml_model = ct.models.MLModel('path/to/your/model.mlpackage')

   # Inspect metadata
   print(coreml_model.user_defined_metadata)
   ```

If the issue persists, please provide a more detailed minimum reproducible example, including the specific steps and code used for training, exporting, and testing the model. This will help us reproduce the issue on our end and investigate further. You can refer to our [minimum reproducible example guide](https://docs.ultralytics.com/help/minimum_reproducible_example) for more details.

Thank you for your cooperation and patience. We're here to help you get this resolved! 😊

**Comment by tensorturtle** [comment]
Hi all, I have noticed the same issue. But before I explain, I think @glenn-jocher should reconsider the use of AI-generated responses in issues like this. If human @glenn-jocher could look over the responses that are given to @tardoe 's messages, the repetitiveness basically turns this github issues thread into spam.

I also had the issue where a custom fine-tuned model, when exported to CoreML, showed either no results or completely wrong results in Xcode.

Right now my workaround is using old versions: 
+ ultralytics==8.0.124 
+ coremltools==6.2

Those versions just happen to be the version that I can confirm working, but are not the 'last' versions where this bug doesn't exit. I will report back once I find that point- the point where the bug was introduced. We can look at the code then.

**Comment by tardoe** [comment]
@tensorturtle i tend to agree, a couple LLM responses is fine but after that it’s just going round in circles. I ended up hopping into the discord and spoke to humans instead. 

TLDR: I concluded that my multi-class model wasn’t trained well enough for CoreML. A two-class model with hundreds of training images per class (vs 50-60 per class previously) worked well enough.

I’ll have to try with your versions and report back.

**Comment by tensorturtle** [comment]
Hi @tardoe I spent some time figuring it out last night:

TLDR: The fix, which works on latest version 8.2.72, is to use `format="mlmodel"` instead of `format="coreml"`.

+ The bug was introduced when the ultralytics codebase upgraded the `coremltools` version from `6.2` to `7.0b1`, which was from `8.0.149` to `8.0.150`.
+ `coremltools==7.0` exports a `.mlpackage`, which is supposed to be an upgrade from `6.2`'s `.mlmodel`. The `.mlpackage`, when imported into XCode, doesn't work. The `.mlmodel` does work.
+ I haven't had the time to try to actually debug the error in the ultralytics code, which is here: https://github.com/ultralytics/ultralytics/blob/bea4c932783f6ab80cce91e00abe38d45c575531/ultralytics/engine/exporter.py#L605
+ Thankfully, even on the current version `8.2.72`, it is possible to use the 'legacy' `coremltools==6.2` which exports `.mlmodel` by passing in `format="mlmodel"`.

I would consider re-opening this issue so that the ultralytics team can be aware of this bug.

**Comment by tardoe** [comment]
Let me try this myself - I'll report back!

**Comment by pderrenger** [comment]
@tardoe thank you for your patience. Please try using `format="mlmodel"` instead of `format="coreml"` when exporting your model. This should generate a `.mlmodel` file, which has been reported to work correctly in Xcode. If the issue persists, ensure you are using the latest versions of `ultralytics` and `coremltools`. Let us know how it goes!

**Comment by tensorturtle** [comment]
@pderrenger Can human @pderrenger look at how silly this AI generated comment is and stop it?

**Comment by pderrenger** [comment]
Thank you for your feedback. We understand the importance of addressing issues accurately and efficiently. Please ensure you are using the latest versions of `ultralytics` and `coremltools`. If the problem persists, try exporting your model using `format="mlmodel"` instead of `format="coreml"`. This should generate a `.mlmodel` file, which has been reported to work correctly in Xcode. If you encounter further issues, feel free to share more details.

**Comment by korotkinmikhail** [comment]
@tardoe @tensorturtle  Hi, I'm currently implementing my yollo v8 model to my app and honestly I'm facing with the same issues during model convertation. I'm really like your analysis from this tred, so - **could you please confirm that you were manage to run your model inside the app using ultralytics==8.0.124 and coremltools==6.2?** It could save me a lot of time)
My model size is 47mb and it has 170 classes (box losses only 10% and it worls fine inside colab and on ios simulator. but not within the app in run time).
If your's analysis leads to dead end it would help me to change the app architecure (I assume I can put my model to S3 and send external requests from the app throught heroku(?) to receive the response I need)

**Comment by korotkinmikhail** [comment]
I've check it by myself - It works. My model now working on coremltools==6.2. @tensorturtle  Thank you so match - me and my team spent 3 day trying to understand what's going on wrong and also huge thanks to @tardoe for starting this tred

**Comment by tensorturtle** [comment]
@korotkinmikhail Good news for you, it turns out we can use the current version `ultralytics==8.2.73` with `coremltools==6.2` ! The solution is to use `yolo export format="mlmodel" ... ` instead of `yolo export format="coreml" ...`.

The bug:
The documentation says to use `yolo export format="coreml" ...` for exporting to iOS. When the resulting `.mlpackage` is copied and imported into XCode and the inference result previewed on a sample image, either no results or nonsense results are shown.

When it was introduced:
From exhaustive testing, I figured out that the bug was introduced in `ultralytics` version 8.0.150, when the `coremltools` was upgraded from 6.2 to 7.0b1. Version 6.2's output was a `.mlmodel`, whereas version 7.0b1 and above outputs `.mlpackage`. Somewhere in the logic of the `.mlpackage` creation something goes wrong.

Fix / Workaround:
I noticed that setting the format argument to "mlmodel" instead of "coreml" uses the legacy .mlmodel format by limiting maximum coremltools version to 6.2. Since we established earlier that it's the upgrade to coremltools 7.0 that caused the problem, we can now run the latest ultralytics version (8.2.73) with this legacy option. `coremltools` 6.2 will output `.mlmodel`, which is still recognized and compatible with current versions of XCode (15.1) and iOS (17.5.1). There are options inside XCode to convert it to a `.mlpackage`. As of this writing, the underlying `mlpackage` creation bug is not actually resolved.

Reference:
I had to read the actual code to find out this legacy option existed. `engine/exporter.py`:  https://github.com/ultralytics/ultralytics/blob/bea4c932783f6ab80cce91e00abe38d45c575531/ultralytics/engine/exporter.py#L605

**Comment by hanrw** [comment]
Issue similar to the one encountered with the format=coreml, resulting in inconsistent predictions compared to the original PyTorch model:

Using format=coreml leads to incorrect predictions.
Using format=mlmodel results in accurate predictions.

**Comment by glenn-jocher** [comment]
@hanrw thank you for sharing your experience. It seems that using `format=mlmodel` is currently providing more consistent results. We recommend continuing with this format for accurate predictions while we investigate the issue with `format=coreml`.

**Comment by hanrw** [comment]
> @hanrw thank you for sharing your experience. It seems that using `format=mlmodel` is currently providing more consistent results. We recommend continuing with this format for accurate predictions while we investigate the issue with `format=coreml`.

i guess it's related to https://github.com/ultralytics/hub/issues/907#issuecomment-2453591753
![image](https://github.com/user-attachments/assets/d19b3a18-14ff-4b96-8ee8-853f50faae5d)

**Comment by glenn-jocher** [comment]
@hanrw thank you for pointing that out. We'll look into the issue further to ensure consistent results across all formats.

**Comment by GuillaumeGenest** [comment]
Hi

I'm facing the same issue. I trained a YOLO11 model with only one class. The model performed well during training, validation, and testing.
                                                                                            
Validating runs/detect/train/weights/best.pt...                                                                                                                                                       
Ultralytics 8.3.49 🚀 Python-3.9.6 torch-2.2.2 CPU (Intel Core(TM) i3-8100B 3.60GHz)                                                                                                                  
YOLO11n summary (fused): 238 layers, 2,582,347 parameters, 0 gradients, 6.3 GFLOPs                                                                                                                    
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:09<00:00,  3.29s/it]                                                               
                   all         79       1229      0.955      0.908      0.971      0.861                                                                                                              
Speed: 3.1ms preprocess, 111.1ms inference, 0.0ms loss, 0.8ms postprocess per image                                                                                                                   

If you export the model with format=coreml and nms=true, it generates a .mlpackage file. In Xcode, this allows the preview to analyze images and detect objects correctly. However, it does not work when using **model.prediction(input: input)** from the class or with **VNCoreMLRequest(model: YOLOModel)**.

To resolve this issue, you need to export the model using format=mlmodel with coremltools==6.2. This ensures compatibility with both methods, and everything works as expected.

thanks @tardoe
